homeassistant/components/sonarr/sensor.py

```diff
@@ -86,164 +97,71 @@ async def async_setup_entry(
     async_add_entities: AddEntitiesCallback,
 ) -> None:
     """Set up Sonarr sensors based on a config entry."""
-    sonarr: SonarrClient = hass.data[DOMAIN][entry.entry_id][DATA_SONARR]
-    host_config: PyArrHostConfiguration = hass.data[DOMAIN][entry.entry_id][
-        DATA_HOST_CONFIG
+    coordinators: dict[str, SonarrDataUpdateCoordinator] = hass.data[DOMAIN][
+        entry.entry_id
     ]
-    system_status: SystemStatus = hass.data[DOMAIN][entry.entry_id][DATA_SYSTEM_STATUS]
-    options: dict[str, Any] = dict(entry.options)
-
-    entities = [
+    async_add_entities(
         SonarrSensor(
-            sonarr,
-            host_config,
-            system_status,
-            entry.entry_id,
+            coordinators[coordinator_type],
             description,
-            options,
         )
-        for description in SENSOR_TYPES
-    ]
-
-    async_add_entities(entities, True)
-
-
-def sonarr_exception_handler(
-    func: Callable[Concatenate[_SonarrSensorT, _P], Awaitable[None]]
-) -> Callable[Concatenate[_SonarrSensorT, _P], Coroutine[Any, Any, None]]:
-    """Decorate Sonarr calls to handle Sonarr exceptions.
-
-    A decorator that wraps the passed in function, catches Sonarr errors,
-    and handles the availability of the entity.
-    """
-
-    @wraps(func)
-    async def wrapper(
-        self: _SonarrSensorT, *args: _P.args, **kwargs: _P.kwargs
-    ) -> None:
-        try:
-            await func(self, *args, **kwargs)
-            self.last_update_success = True
-        except ArrConnectionException as error:
-            if self.last_update_success:
-                _LOGGER.error("Error communicating with API: %s", error)
-            self.last_update_success = False
-        except ArrException as error:
-            if self.last_update_success:
-                _LOGGER.error("Invalid response from API: %s", error)
-                self.last_update_success = False
-
-    return wrapper
+        for coordinator_type, description in SENSOR_TYPES.items()
+    )
 
 
 class SonarrSensor(SonarrEntity, SensorEntity):
```

Comment:
`SonarEntity` is also a Generic.

```suggestion
class SonarrSensor(SonarrEntity[SonarDataT], SensorEntity):
```